[SPARK-28948][SQL] Support passing all Table metadata in TableProvider

cloud-fan opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

Currently Data Source V2 has 2 major use cases:
1. users plug in a custom catalog, which is tightly coupled with its own data. For example, users can plug in a cassandra catalog, and use Spark to read/write cassandra tables directly.
2. users read/write the external data as a table directly via `DataFrameReader/Writer`, or register it as a table in Spark.

Use case 1 is newly introduced in the master branch, which greatly improves the user experience when interacting with external storage systems that have catalogs, e.g. cassandra, JDBC, etc.

Use case 2 is the main use case of Data Source V1, which works well if the external storage system doesn't have a catalog, e.g. parquet files on S3.

However, use case 2 is not well supported. For example
```
class MyTableProvider extends TableProvider ...
sql("CREATE TABLE t USING com.abc.MyTableProvider")
```
This fails with `AnalysisException: com.abc.MyTableProvider is not a valid Spark SQL Data Source`. The session catalog always treats table provider as v1 source.

To support it, this PR updates `TableProvider#getTable` to accept additional table metadata info. The expected behaviors are defined in https://docs.google.com/document/d/1oaS0eIVL1WsCjr4CqIpRv6CGkS5EoMQrngn3FsY1d-Q/edit?usp=sharing


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Make Data Source V2 supports the use case that is supported by Data Source V1.

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->
Yes, it's a new feature

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
a new test suite

SparkQA commented:
**[Test build #110014 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110014/testReport)** for PR 25651 at commit [`7fd7d23`](https://github.com/apache/spark/commit/7fd7d23113ad46bf002d1dbaf518a474809d7273).

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/sources/v2/TableProvider.java:
```diff
@@ -44,18 +45,28 @@
    * @throws UnsupportedOperationException
    */
-  default Table getTable(CaseInsensitiveStringMap options, StructType schema) {
+  default Table getTable(
```
I'll refine the classdoc of this interface, after we reach an agreement of the proposal.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15034/
Test PASSed.

cloud-fan commented:
cc @brkyvz @rdblue @jose-torres @gengliangwang @gatorsmile

SparkQA commented:
**[Test build #110014 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110014/testReport)** for PR 25651 at commit [`7fd7d23`](https://github.com/apache/spark/commit/7fd7d23113ad46bf002d1dbaf518a474809d7273).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110014/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15045/
Test PASSed.

SparkQA commented:
**[Test build #110027 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110027/testReport)** for PR 25651 at commit [`0da5453`](https://github.com/apache/spark/commit/0da5453549a53cab720533b212b24cdd83e4640b).

SparkQA commented:
**[Test build #110027 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110027/testReport)** for PR 25651 at commit [`0da5453`](https://github.com/apache/spark/commit/0da5453549a53cab720533b212b24cdd83e4640b).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110027/
Test FAILed.

gengliangwang commented:
retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15053/
Test PASSed.

SparkQA commented:
**[Test build #110035 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110035/testReport)** for PR 25651 at commit [`0da5453`](https://github.com/apache/spark/commit/0da5453549a53cab720533b212b24cdd83e4640b).

SparkQA commented:
**[Test build #110035 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110035/testReport)** for PR 25651 at commit [`0da5453`](https://github.com/apache/spark/commit/0da5453549a53cab720533b212b24cdd83e4640b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110035/
Test FAILed.

gengliangwang commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:
```diff
@@ -180,7 +270,29 @@ class V2SessionCatalog(sessionState: SessionState) extends TableCatalog {
 }
 
-private[sql] object V2SessionCatalog {
+case class SchemaChangedException(
```
How about just `SchemaException`

gengliangwang commented:
+1 with the proposal.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:
```diff
@@ -216,7 +216,7 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       val dsOptions = new CaseInsensitiveStringMap(finalOptions.asJava)
       val table = userSpecifiedSchema match {
-        case Some(schema) => provider.getTable(dsOptions, schema)
+        case Some(schema) => provider.getTable(dsOptions, schema, Array.empty)
```
This isn't correct. The `DataFrameReader` does not know that the table is unpartitioned.

brkyvz replied:
yeah, this is one of the most annoying and confusing behaviors of DataSourceV1. Being able to provide a schema but not partitioning information, which leads to minutes of partition schema inference.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanSupportCheck.scala:
```diff
@@ -20,17 +20,20 @@ package org.apache.spark.sql.execution.datasources.v2
     plan.foreach {
+      case r: DataSourceV2Relation if !r.table.supports(BATCH_READ) =>
+        throw new AnalysisException(
+          s"Table ${r.table.name()} does not support batch scan.")
```
Why was this change included here and not in a separate PR?

cloud-fan replied:
I've created a separate PR to fix it: https://github.com/apache/spark/pull/25679

rdblue commented on external/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroDataSourceV2.scala:
```diff
@@ -35,7 +36,10 @@ class AvroDataSourceV2 extends FileDataSourceV2 {
+  override def getTable(
+      options: CaseInsensitiveStringMap,
+      schema: StructType,
+      partitions: Array[Transform]): Table = {
```
Why is this not identical to the metadata passed to `createTable`? Is there a reason not to pass the table properties as well as the read options?

brkyvz replied:
I agree, this should look similar to `createTable`

rdblue replied:
This interface should pass the table properties. There is no need to pass read or write options at this point, unless they can't be separated from table properties (as in the `DataFrameReader` case). The read options and write options should be passed to the logical plan -- this is added in #25681: https://github.com/apache/spark/pull/25681/files#diff-94fbd986b04087223f53697d4b6cab24R275

I propose passing table properties as a string map (java.util) through this interface. When the properties come from the metastore, then this is fine. When the properties come from `DataFrameReader.option` (or the write equivalent) then the original case sensitive map should be passed. Then the read options should additionally be passed to the correct plan node so that the physical plan can push them into the scan or the write.

jose-torres commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:
```diff
@@ -63,6 +63,61 @@ class V2SessionCatalog(sessionState: SessionState) extends TableCatalog {
+      val partitions = new mutable.ArrayBuffer[Transform]()
+
+      v1Table.partitionColumnNames.foreach { col =>
+        partitions += LogicalExpressions.identity(col)
```
Nit: This parses the column's name as a multi-part identifier, which is subtly incorrect. (It'll cause issues if the column name contains special characters like ':'.)

rdblue replied:
Aren't column names with special characters supported? I thought you could escape any identifier using back-ticks.

rdblue commented:
@cloud-fan, can you update the PR title and description? The `USING` clause is not the problem. That is passed to all catalogs. The problem is that generic catalogs can't pass table information to `Table` instances created by `TableProvider`. I think the title should be "Support passing all Table metadata in TableProvider".

I think that clarifying the problem statement will also help clean up the proposed changes. For example, this passes some -- but not all -- `Table` metadata. It should probably pass all of the fields needed to create a `Table` instance that behaves like `UnresolvedTable`.

AmplabJenkins commented:
Build finished. Test PASSed.

cloud-fan committed:
Support passing all Table metadata in TableProvider

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16034/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableProvider.java:
```diff
@@ -36,26 +40,21 @@
+  Table getTable(
+      Optional<StructType> schema,
+      Optional<Transform[]> partitions,
+      Map<String, String> properties);
```
I'd like to discuss how the API should look like. The current use cases include
1. users only specify options, implementation needs to infer schema/partitioning
2. users specify options and schema, implementation needs to infer partitioning
3. users specify all the things.

Shall we create 3 methods or just create one single method like this?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:
```diff
@@ -66,7 +105,11 @@ class V2SessionCatalog(catalog: SessionCatalog, conf: SQLConf)
+      throw new NoSuchTableException(ident)
+    }
+
+    tryResolveTableProvider(V1Table(catalogTable))
```
This is the major change of this PR.

SparkQA commented:
**[Test build #110919 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110919/testReport)** for PR 25651 at commit [`2333585`](https://github.com/apache/spark/commit/233358518259fdb3045dbffbd19cad8dd11b9c32).

SparkQA commented:
**[Test build #110919 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110919/testReport)** for PR 25651 at commit [`2333585`](https://github.com/apache/spark/commit/233358518259fdb3045dbffbd19cad8dd11b9c32).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110919/
Test FAILed.

SparkQA commented:
**[Test build #111290 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111290/testReport)** for PR 25651 at commit [`40e2894`](https://github.com/apache/spark/commit/40e289406142c09c5c7a7fed21e4523c2f762ab0).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16349/
Test PASSed.

SparkQA commented:
**[Test build #111290 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111290/testReport)** for PR 25651 at commit [`40e2894`](https://github.com/apache/spark/commit/40e289406142c09c5c7a7fed21e4523c2f762ab0).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CatalogExtensionForTableProvider extends DelegatingCatalogExtension `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111290/
Test FAILed.

SparkQA commented:
**[Test build #111293 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111293/testReport)** for PR 25651 at commit [`3a6d13d`](https://github.com/apache/spark/commit/3a6d13dd600c661390bcd7c85d7e663cdcc2c521).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16350/
Test PASSed.

SparkQA commented:
**[Test build #111293 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111293/testReport)** for PR 25651 at commit [`3a6d13d`](https://github.com/apache/spark/commit/3a6d13dd600c661390bcd7c85d7e663cdcc2c521).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CatalogExtensionForTableProvider extends DelegatingCatalogExtension `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111293/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16355/
Test FAILed.

SparkQA commented:
**[Test build #111298 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111298/testReport)** for PR 25651 at commit [`0f9faca`](https://github.com/apache/spark/commit/0f9faca3c796768c4b08c61c5eab4a011c7c7b40).

SparkQA commented:
**[Test build #111298 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111298/testReport)** for PR 25651 at commit [`0f9faca`](https://github.com/apache/spark/commit/0f9faca3c796768c4b08c61c5eab4a011c7c7b40).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CatalogExtensionForTableProvider extends DelegatingCatalogExtension `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111298/
Test FAILed.

rdblue commented:
@cloud-fan, thanks for working on this. I plan to review it tomorrow. Looks like this is huge and touches about 50 files. Is there a way to make it smaller?

cloud-fan commented:
@rdblue yea it's possible. In this PR, I try to adopt your suggestion to make it clear that `TableProvider.getTable` should take all the table metadata, so the method signature becomes
```
def getTable(schema: StructType, partitions: Array[Transform], properties: Map[String, String])
```

`TableProvider` has another `getTable` method which needs to infer schema/partitioning, and previously the method signature was
```
def getTable(options: CaseInsensitiveStringMap)
```

To make it consistent, I change it to use `properties: Map[String, String]`, also rename it to `loadTable` since we need to touch many files anyway.

We can still keep the old method signature with a TODO to change it later, so that this PR can be much smaller.

cloud-fan committed:
make the change smaller

SparkQA commented:
**[Test build #111354 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111354/testReport)** for PR 25651 at commit [`2bf1c59`](https://github.com/apache/spark/commit/2bf1c59fb18ee40143c87472b2d3323a51c238ce).

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16400/
Test PASSed.

SparkQA commented:
**[Test build #111357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111357/testReport)** for PR 25651 at commit [`cfc2264`](https://github.com/apache/spark/commit/cfc2264db055e61ab871f96efc84b19f2d6ba68e).

cloud-fan commented:
I've pushed a new commit to keep `def getTable(options: CaseInsensitiveStringMap)` unchanged, with a TODO to update to `def getTable(properties: Map<String, String>)`. This avoids 20 changed files and makes this PR much easier to review.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16403/
Test PASSed.

SparkQA commented:
**[Test build #111354 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111354/testReport)** for PR 25651 at commit [`2bf1c59`](https://github.com/apache/spark/commit/2bf1c59fb18ee40143c87472b2d3323a51c238ce).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `class RateStreamTable(`

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111354/
Test FAILed.

SparkQA commented:
**[Test build #111357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111357/testReport)** for PR 25651 at commit [`cfc2264`](https://github.com/apache/spark/commit/cfc2264db055e61ab871f96efc84b19f2d6ba68e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RateStreamTable(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111357/
Test FAILed.

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala:
```diff
@@ -31,14 +31,15 @@ import org.apache.spark.sql.types.StructType
  * An implementation of catalog v2 `Table` to expose v1 table metadata.
  */
 private[sql] case class V1Table(v1Table: CatalogTable) extends Table {
+  assert(v1Table.provider.isDefined)
```
Why is this guaranteed? What about Hive tables, will those have provider defined?

cloud-fan replied:
Only view has no provider. So this is guaranteed. Please see the parameter doc of `CatalogTable#provider`.

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala:
```diff
@@ -77,16 +77,15 @@ class CatalogManager(
+      case NonFatal(_) =>
+        logError(
+          "Fail to instantiate the custom v2 session catalog: " +
+            conf.getConfString(CatalogManager.SESSION_CATALOG_NAME))
```
Why was this error message changed? The previous one that included the implementation class seems more helpful to me.

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala:
```diff
@@ -51,20 +52,18 @@ private[sql] case class V1Table(v1Table: CatalogTable) extends Table {
-
-  lazy val options: Map[String, String] = {
-    v1Table.storage.locationUri match {
+  override lazy val properties: util.Map[String, String] = {
```
If options are to be included in properties, I think that we should prefix them with `option.` so that we can recover them. I suggested this to Xin, who opened SPARK-28555 and was going to work on it, but I haven't seen a PR for it yet.

The idea is basically to be able to keep options and properties separate when passing through the v2 interface that only supports properties.

Since this PR updates V1Table to combine them, maybe we should include the change in this PR. What do you think?

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala:
```diff
@@ -84,5 +83,5 @@ private[sql] case class V1Table(v1Table: CatalogTable) extends Table {
   override def capabilities: util.Set[TableCapability] = new util.HashSet[TableCapability]()
 
-  override def toString: String = s"UnresolvedTable($name)"
+  override def toString: String = s"V1Table($name)"
```
+1

Good catch, this would have been confusing.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CatalogExtensionForTableProvider.scala:
```diff
@@ -0,0 +1,98 @@
+
+  private val conf = SQLConf.get
+
+  override def loadTable(ident: Identifier): Table = {
```
Why is this written as a catalog extension? I think this should be the default behavior of the v2 wrapper for Spark's catalog. Why use the extension in interface for this?

cloud-fan replied:
I did struggle for it many times. If we put this logic in `V2SessionCatalog`, then the catalog extension API becomes less useful. If a user creates a `CustomExtension` and tries to overwrite the `loadTable` method, then there is not much he can do as the table provider resolving is already done.

rdblue replied:
This should be part of the built-in implementation, `V2SessionCatalog`, not an extension. The extension API has all of the same information -- the table identifier, table schema, partitioning, and properties. It can still return a different instance of `Table` if it needs to.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CatalogExtensionForTableProvider.scala:
```diff
@@ -0,0 +1,98 @@
+      // If `CREATE TABLE ... USING` does not specify table metadata, get the table metadata from
+      // data source first.
+      val table = maybeProvider.get.getTable(new CaseInsensitiveStringMap(properties))
+      table.schema() -> table.partitioning()
```
This block isn't correct. If the table exists, then this must throw `TableAlreadyExistsException`: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableCatalog.java#L99

Create table is not allowed to return an existing table and override partitioning.

cloud-fan replied:
There are 2 "table" in this context:
1. the table entry in Spark metastore
2. the table returned by `TableProvider`

For example, `CREATE TABLE abc USING jdbc OPTIONS(table='xyz')`, `abc` is the table entry in Spark metastore, `xyz` is the table returned by `TableProvider`. The table entry in Spark metastore is simply a link to the table returned by `TableProvider`. So the CREATE TABLE here is not to create a table in JDBC, but to create a table in Spark metastore that links to the JDBC table.

rdblue replied:
Okay, so this create table uses the schema and partitioning from the external table because it wasn't specified in the DDL create statement. Sounds reasonable.

I don't think this implementation is correct. There are two cases: the schema and/or partitioning were not specified and are empty, or they were specified and need to be validated. I think that this should load the external table and either override (if empty) or validate if non-empty.

Looks like we didn't think of this case in the API, but we should add `IllegalArgumentException` for when the external table schema or partitioning doesn't match the incoming settings.

When the `inferSchema` and `inferPartitioning` methods are added, we will need to update this to use those methods instead of loading the table and getting its schema and partitioning. After those methods are added, the table implementation must always report the schema and partitioning that were passed in to build the table.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CatalogExtensionForTableProvider.scala:
```diff
@@ -0,0 +1,98 @@
+    super.createTable(ident, actualSchema, actualPartitioning, properties)
+    // call `loadTable` to make sure the schema/partitioning specified in `CREATE TABLE ... USING`
+    // matches the actual data schema/partitioning. If error happens during table loading, drop
+    // the table.
```
Why would this happen?

cloud-fan replied:
e.g. `CREATE TABLE t(i int) USING jdbc OPTIONS (table=t2)`. It's possible that the JDBC table `t2` has a different schema.

rdblue replied:
This should be validated above when the table is loaded, not by loading the table as second time.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CatalogExtensionForTableProvider.scala:
```diff
@@ -0,0 +1,98 @@
+    assert(providerName != null)
+    DataSource.lookupDataSourceV2(providerName, conf).map {
+      // TODO: support file source v2 in CREATE TABLE USING.
+      case _: FileDataSourceV2 => table
```
Why is file source handled here? File source should use `SupportsSpecifiedSchemaPartitioning.getTable`.

cloud-fan replied:
As you already found out in https://github.com/apache/spark/pull/25651#discussion_r328336988

File source v2 can't take the partitioning from metastore because the `TableProvider` API was incompleted before. I don't want to fix file source v2 in this PR, so I simply ignore it here.

rdblue replied:
Okay, so this will be removed in the commit that updates file sources?

cloud-fan replied:
yup

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala:
```diff
@@ -57,4 +60,21 @@ private[sql] object DataSourceV2Utils extends Logging {
+    provider match {
+      case s: SupportsSpecifiedSchemaPartitioning =>
+        // TODO: `DataFrameReader`/`DataStreamReader` should have an API to set user-specified
+        //       partitioning.
```
This doesn't make sense to me. Partitioning should be passed here. That's what this method is for.

cloud-fan replied:
But there is no API to specify the partitioning here, neither `DataFrameReader` or `DataStreamReader`.

rdblue replied:
If we are fixing the sources later, maybe we can remove this as well and fix it in the follow-up when the partition inference method is done.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:
```diff
@@ -42,7 +42,7 @@ class V2SessionCatalog(catalog: SessionCatalog, conf: SQLConf)
   import V2SessionCatalog._
 
-  override val defaultNamespace: Array[String] = Array("default")
+  override def defaultNamespace: Array[String] = Array("default")
```
Why did this change?

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:
```diff
@@ -121,41 +133,37 @@ class V2SessionCatalog(catalog: SessionCatalog, conf: SQLConf)
-    loadTable(oldIdent)
+    if (!tableExists(oldIdent)) {
+      throw new NoSuchTableException(oldIdent)
+    }
```
These changes don't seem related, but look okay to me.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVDataSourceV2.scala:
```diff
@@ -35,9 +38,12 @@ class CSVDataSourceV2 extends FileDataSourceV2 {
+    val paths = getPaths(properties)
     val tableName = getTableName(paths)
-    CSVTable(tableName, sparkSession, options, paths, Some(schema), fallbackFileFormat)
+    CSVTable(tableName, sparkSession, properties, paths, Some(schema), fallbackFileFormat)
```
`partitions` is ignored in all of these implementations. That looks like a bug to me.

cloud-fan replied:
We shouldn't make this PR bigger and supports partitioning in file source. For now let's explicitly say that file source doesn't support user-specified partitioning: https://github.com/apache/spark/pull/25651/files#diff-3d5fde6e98a2856ba1b00ce6f172c4a8R72

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVDataSourceV2.scala:
```diff
@@ -35,9 +38,12 @@ class CSVDataSourceV2 extends FileDataSourceV2 {
-    val paths = getPaths(options)
+  override def getTable(
+      schema: StructType,
+      partitions: Array[Transform],
```
The name used for `Array[Transform]` in other places is `partitioning` to distinguish the transforms from actual partition locations. Can you update to use that convention instead?

cloud-fan replied:
shall we also update `TableCatalog.createTable`?

rdblue replied:
Oh, I didn't realize that there was precedent for this. I'd rather not make a change to existing code just to rename, so I guess this is okay. I'd prefer to rename here at least, but this is minor.

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableProvider.java:
```diff
@@ -36,26 +35,12 @@
    * @param options the user-specified options that can identify a table, e.g. file path, Kafka
    *                topic name, etc. It's an immutable case-insensitive string-to-string map.
    */
+  // TODO: this should take a Map<String, String> as table properties.
```
If this method is left, then all of the changes to file sources can be avoided in this commit. That would make it considerably smaller.

rdblue replied:
Another option for avoiding this is to separate the schema and partition inference from the `getTable` method. In that case, `TableProvider` would expose `inferSchema(CaseInsensitiveStringMap)` and `inferPartitioning(CaseInsensitiveStringMap)`. Then a single `getTable` call could be used.

```java
interface TableProvider {
  StructType inferSchema(CaseInsensitiveStringMap options);
  Transform[] inferPartitioning(CaseInsensitiveStringMap options);
  Table buildTable(StructType schema, Transform[] partitioning, Map<String, String> properties);
}
```

What do you think?

cloud-fan replied:
SGTM. Shall we do it in this PR? We need to update all the `TableProvider` implementations, i.e. file source v2, streaming source v2, testing v2 sources.

rdblue replied:
No, I think we should revert the changes to all the sources in this commit and continue to use the original method. In this PR, let's add the method to build a table and add support to the v2 session catalog. We can change to infer before calling that method in a follow-up.

cloud-fan replied:
I did a quick try, and found one problem: it's very likely that a data source needs to infer schema and partition together, e.g. file source. If we have 2 separated method `inferSchema` and `inferPartitioning`, it may force the data source to do expensive inference twice.

rdblue replied:
Can't we do `inferPartitioning` first and then pass that into `inferSchema`?

cloud-fan replied:
Then API is limited to the assumption that schema inference depends on partition inference. This is true for file source, but I'm not sure if it's true to all data sources.

Shall we stick to the current proposal that has 2 `getTable` methods?

cloud-fan replied:
It turns out that this is more complicated than I thought.

For file source, we need to list the files to scan. If we need to infer the schema/partitioning, we will list the files a little earlier, and reuse the listed files when scan.

That said, we need to either make `TableProvider` stateful, so that file source can keep the listed files after schema inference. Or we need to re-design the API so that it can carry states after schema inference.

Maybe we should move on with 2 `getTable` methods as it is in this PR.

rdblue replied:
A TableProvider could use a static cache. We do this for our Hive client pool in Iceberg.

I don't think that partition inference needs to scan the entire file system tree. That seems needlessly expensive to me, when mixed depth partitions are not allowed. Inference should find the first live data file and use its path to determine the partitioning.

In any case, I think that the plan to add `inferSchema` and `inferPartitioning` is fine, it is just fixing the implementation that needs to be done. Schema and partitioning inference is known to be expensive, so I'm not sure how much sense it makes to try to over-optimize here.

cloud-fan replied:
> I don't think that partition inference needs to scan the entire file system tree.

Spark needs to do it to get all the partition values and infer the schema. This is an existing feature that Spark can infer a common type for partition values with different types. The same applies to schema inference as well. Spark can read parquet files of different but compatible schema, so Spark must read all the files to infer the schema.

Can you share more about the static cache? Do you mean a global cache that maps a directory to its listed files?

rdblue replied:
My point was that Spark needs to infer the partitioning of the table, not exhaustively list directories. This can be done more quickly than in the current implementation, by listing all files later and just getting the directory structure for `inferPartititoning`.

The static cache I'm talking about is a cache of metastore connections, not files. In this case, you could build your file list for a location and cache that for some period of time, using it for partition and schema inference, as well as for the `FileIndex` in the table you created. Caching would also help consistency because the same files would be in all versions of the table loaded within some period of time (and could be refreshed, of course). But, these concerns shouldn't affect the API.

rdblue commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsSpecifiedSchemaPartitioning.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * A mix-in interface for {@link TableProvider}. Data sources can implement this interface to
+ * return a table instance with specified schema and partitioning, so that they may avoid expensive
+ * schema/partitioning inference.
```
Why is this a mix-in? This is required for a `TableProvider` to be used by any generic catalog implementation, so I think it should be a requirement of `TableProvider` itself.

cloud-fan committed:
address comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16880/
Test PASSed.

SparkQA commented:
**[Test build #111902 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111902/testReport)** for PR 25651 at commit [`f2888c8`](https://github.com/apache/spark/commit/f2888c8859e62ca0109055c31fbf80bf4ecab2a8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111902/
Test FAILed.

SparkQA commented:
**[Test build #111903 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111903/testReport)** for PR 25651 at commit [`5226bb2`](https://github.com/apache/spark/commit/5226bb27389139b15dca8a41973dacf0103fa18b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16883/
Test PASSed.

SparkQA commented:
**[Test build #111905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111905/testReport)** for PR 25651 at commit [`1124b47`](https://github.com/apache/spark/commit/1124b47a4cb943336b9d431d91121beb6702c699).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V1Table.scala:
```diff
@@ -51,38 +52,36 @@ private[sql] case class V1Table(v1Table: CatalogTable) extends Table {
   }
 
-  override lazy val properties: util.Map[String, String] = v1Table.properties.asJava
+  override lazy val properties: util.Map[String, String] = catalogTable.properties.asJava
```
I leave the options/properties unchanged here, but we need to figure it out later. Currently there are 2 directions:
1. We have table options and table properties. Table options are special table properties with "option." prefix in its name. Table options will be extracted and used as scan/write options.
2. We only have table properties. The OPTIONS clause in CREATE TABLE should be the same as the TBLPROPERTIES clause.

We can have more discussion about it later.

rdblue replied:
V2 has only table properties, so I think the question is whether we want to mix options into those table properties directly, or whether we want to prefix them so they can be recovered. I'm in favor of being able to recover them so we can constract the catalog table as it would be in the v1 path.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16887/
Test PASSed.

cloud-fan commented:
Hi @rdblue , I think I have addressed all your comments except https://github.com/apache/spark/pull/25651#discussion_r332095472

Please let me know if you have more concerns on that topic. Thanks!

SparkQA commented:
**[Test build #111903 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111903/testReport)** for PR 25651 at commit [`5226bb2`](https://github.com/apache/spark/commit/5226bb27389139b15dca8a41973dacf0103fa18b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111903/
Test FAILed.

SparkQA commented:
**[Test build #111905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111905/testReport)** for PR 25651 at commit [`1124b47`](https://github.com/apache/spark/commit/1124b47a4cb943336b9d431d91121beb6702c699).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111905/
Test FAILed.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #112044 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112044/testReport)** for PR 25651 at commit [`1124b47`](https://github.com/apache/spark/commit/1124b47a4cb943336b9d431d91121beb6702c699).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17050/
Test PASSed.

SparkQA commented:
**[Test build #112044 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112044/testReport)** for PR 25651 at commit [`1124b47`](https://github.com/apache/spark/commit/1124b47a4cb943336b9d431d91121beb6702c699).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112044/
Test FAILed.

cloud-fan committed:
Merge remote-tracking branch 'origin/master' into dsv2-using

cloud-fan committed:
update

SparkQA commented:
**[Test build #112108 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112108/testReport)** for PR 25651 at commit [`1235d78`](https://github.com/apache/spark/commit/1235d78badc358db0ed586f9419a2bf766a4e22b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17104/
Test PASSed.

SparkQA commented:
**[Test build #112108 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112108/testReport)** for PR 25651 at commit [`1235d78`](https://github.com/apache/spark/commit/1235d78badc358db0ed586f9419a2bf766a4e22b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112108/
Test FAILed.

rdblue commented:
I should have time to review this on Thursday if it is ready. Until then, I commented on some of the open threads.

cloud-fan commented:
@rdblue , when I try to have separated methods `inferSchema` and `inferPartitioning`, problems keep popping up (mostly from file source v2).

The problems hit so far:
1. `inferSchema`/`inferPartitioning` need to list files, and we should only do file listing once when we scan a directory without user-specified schema. This can be resolved by using a static cache or simply cache the listed files in the `FileDataSourceV2` instance.
2. when writing to a directory, no schema/partition inference should be done. It looks to me that we need to have 2 separated methods `getTableToRead` and `getTableToWrite`, while `getTableToWrite` does not take schema/partitioning and we don't need to call `inferSchema`/`inferPartitioning`. But this makes the API ugly.

I feel that, the existing API (`getTable` with several overloads) is more flexible and allows implementations to have its special logic. For example, it allows file source to do schema inference lazily, which won't be triggered at all during write.

cloud-fan committed:
fix test

SparkQA commented:
**[Test build #112405 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112405/testReport)** for PR 25651 at commit [`cfbe0a7`](https://github.com/apache/spark/commit/cfbe0a75f80e88d4a5831785d05fb9b708c5ada3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17376/
Test PASSed.

SparkQA commented:
**[Test build #112405 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112405/testReport)** for PR 25651 at commit [`cfbe0a7`](https://github.com/apache/spark/commit/cfbe0a75f80e88d4a5831785d05fb9b708c5ada3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112405/
Test FAILed.

rdblue commented:
> inferSchema/inferPartitioning need to list files, and we should only do file listing once when we scan a directory without user-specified schema

Why should this internal concern of one source affect the API? Partitioning inference does not require listing all the files in a table, and I doubt that it is a good idea to do that for schema inference either. If a table is small, it doesn't matter if this work is done twice (before it is fixed); and if a table is really large, then it isn't a good idea to do this for schema inference anyway.

> when writing to a directory, no schema/partition inference should be done.

This statement makes assumptions about the behavior of path-based tables and that behavior hasn't been clearly defined yet. Can you be more specific about the case and how you think path-based tables will behave?

I disagree that no schema or partition inference should be done for writing. Maybe it isn't done today, but if there is existing data, Spark shouldn't allow writing new data that will break a table by using an incompatible schema or partition layout. In that case, we would want to infer the schema and partitioning.

Also, if it isn't necessary to infer schema and partitioning, then this information still needs to be passed to the table. When running a CTAS operation, Spark might be called with `partitionBy`. In that case, if Spark doesn't call `inferPartitioning` then what is the problem?

cloud-fan commented:
I was not trying to define the behavior, but talking about the existing behavior. `df.write.mode("append").parquet("path_with_existing_data")` will not do schema/partition inference, and just append the data even if the schema is incompatible. The following read will fail during schema inference. I think this is a reasonable behavior, as there is no "user-specified schema" in `DataFrameReader` to skip schema inference.

cloud-fan commented:
If you really think that schema or partition inference should be done for writing, we should disable file source v2 by default to not surprise users.

rdblue commented:
> If you really think that schema or partition inference should be done for writing, we should disable file source v2 by default to not surprise users.

I'm not saying that it is what we should do. That should be covered by a design doc for path-based tables. My point is that the claim that it won't be done is not necessarily true and makes assumptions about how these tables will behave.

cloud-fan commented:
Do you mean we should block this PR until we figure out the behavior of path-based tables? This PR simply makes `TableProvider` to accept user-specified partitioning, while keeping the API style and existing file source behavior unchanged. I think we've gone too far about proposing a new API style for `TableProvider`.

I'm OK to adopt the new API style if it doesn't break the existing behavior. But seems now we are unable to keep file source skipping schema/partition inference during write. Shall we discuss the new API style later?

rdblue commented:
> Do you mean we should block this PR until we figure out the behavior of path-based tables?

No, and sorry for the misunderstanding! My point is that your claim that inference won't be used in the write path is not necessarily correct and depends on the behavior we decide for path-based tables.

> But seems now we are unable to keep file source skipping schema/partition inference during write.

I think it's an exaggeration to say "unable". Partition inference in particular can be done much more easily and efficiently than depending on a recursive directory listing to find all data files. Granted, the current implementation would need to change, but do you really think that "unable" is an accurate description?

The problem is that this needs to be decided because it affects the API that will go into Spark 3.0. I think we should go with what we agreed was a good solution for the API -- adding the `inferSchema` and `inferPartitioning` methods -- because I haven't heard a very strong argument against it. Let's talk about this in the next v2 sync to get more opinions.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19832/
Test PASSed.

SparkQA commented:
**[Test build #122766 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122766/testReport)** for PR 25651 at commit [`cfbe0a7`](https://github.com/apache/spark/commit/cfbe0a75f80e88d4a5831785d05fb9b708c5ada3).

SparkQA commented:
**[Test build #122766 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122766/testReport)** for PR 25651 at commit [`cfbe0a7`](https://github.com/apache/spark/commit/cfbe0a75f80e88d4a5831785d05fb9b708c5ada3).
 * This patch **fails build dependency tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122766/
Test FAILed.
